src/librustc_codegen_ironox/type_.rs

```diff
@@ -62,14 +62,32 @@ pub enum LLType {
 }
 
 impl CodegenCx<'ll, 'tcx> {
-    crate fn type_named_struct(&self, name: &str) -> Type {
+    /// Add the specified `LLType` to the vector of `Type`s for this context.
+    ///
+    /// If the type already exists in the type cache, its index in the type
+    /// vector is returned. Otherwise, it is added to the type vector, and to
+    /// the type cache.
+    fn add_type(&self, ll_type: LLType) -> Type {
         let mut types = self.types.borrow_mut();
+        // If ll_type is not in types, it will be inserted at the end of the
+        // types vector.
+        let ty_idx = types.len();
+        // FIXME: don't always clone ll_type.
```

Comment:
Is this a performance issue or a semantic issue?

Reply:
It's a performance issue.

Reply:
OK, then we can ignore that for the time being.